submit-queue/submit-queue.go

```diff
@@ -152,42 +156,43 @@ func (e *e2eTester) runE2ETests(client *github_api.Client, pr *github_api.PullRe
 		return err
 	}
 	if !ok {
-		e.msg(fmt.Sprintf("Status after build is not 'success', skipping PR %d", *pr.Number))
+		e.msg("Status after build is not 'success', skipping PR %d", *pr.Number)
 		return nil
 	}
 	return e.merge(client, org, project, pr)
 }
 
 func (e *e2eTester) merge(client *github_api.Client, org, project string, pr *github_api.PullRequest) error {
-	if !*dryrun {
-		e.msg(fmt.Sprintf("Merging PR: %d", *pr.Number))
-		mergeBody := "Automatic merge from SubmitQueue"
-		if _, _, err := client.Issues.CreateComment(org, project, *pr.Number, &github_api.IssueComment{Body: &mergeBody}); err != nil {
-			glog.Warningf("Failed to create merge comment: %v", err)
-			e.error(err)
-			return err
-		}
-		if _, _, err := client.PullRequests.Merge(org, project, *pr.Number, "Auto commit by PR queue bot"); err != nil {
-			e.error(err)
-			return err
-		}
+	if *dryrun {
+		e.msg("Skipping actual merge because --dry-run is set")
 		return nil
 	}
-	e.msg("Skipping actual merge because --dry-run is set")
+	e.msg("Merging PR: %d", *pr.Number)
+	mergeBody := "Automatic merge from SubmitQueue"
+	if _, _, err := client.Issues.CreateComment(org, project, *pr.Number, &github_api.IssueComment{Body: &mergeBody}); err != nil {
+		e.msg("Failed to create merge comment: %v", err)
+		e.error(err)
+		return err
+	}
+	if _, _, err := client.PullRequests.Merge(org, project, *pr.Number, "Auto commit by PR queue bot"); err != nil {
+		e.msg("Failed to merge PR %d: %v", *pr.Number, err)
+		e.error(err)
+		return err
+	}
 	return nil
 }
 
 func (e *e2eTester) ServeHTTP(res http.ResponseWriter, req *http.Request) {
-	e.Lock()
-	defer e.Unlock()
-	res.Header().Set("Content-type", "application/json")
-	if data, err := json.Marshal(e); err != nil {
-		res.WriteHeader(http.StatusInternalServerError)
-		res.Write([]byte(err.Error()))
-	} else {
-		res.WriteHeader(http.StatusOK)
-		res.Write(data)
-	}
+	e.locked(func() {
+		res.Header().Set("Content-type", "application/json")
+		if data, err := json.MarshalIndent(e, "", "\t"); err != nil {
```

Comment:
it probably doesn't really matter for this app, but don't we only need to lock the `json.MarshalIndent(e...)` bit? the rest of the HTTP stuff shouldn't need to be locked.